### PR TITLE
[WIP] Add checks for particular compound distributions

### PR DESCRIPTION
**Pull request by akash9712**
#### Brief description of what is fixed or changed    
Add function to check if compounding particular distribution returns RVs
whose distribution is known already. As of now, this is only done for
univariate normal distribution with a normal mean.
Example:
```
>>> N1 = Normal('N1', 0, 1)
>>> N2 = Normal('N2', N1, 2)
>>> pspace(N2)
SingleContinuousPSpace(N2, NormalDistribution(0, sqrt(5)))
```

### CODE REVIEW

File: sympy/stats/crv.py [hunk 1]
```diff
@@ -138,12 +139,7 @@ def set(self):
                 "Set of Conditional Domain not Implemented")
 
 
-class ContinuousDistribution(Basic):
```
Comment (Upabjojr) [inline sympy/stats/crv.py]: are you removing the distinction between continuous and discrete distributions?

File: sympy/stats/crv_types.py [hunk 2]
```diff
@@ -845,7 +851,7 @@ def check(rate):
         _value_check(rate > 0, "Rate must be positive.")
 
     def pdf(self, x):
-        return self.rate * exp(-self.rate*x)
+        return  self.rate * exp(-self.rate*x)
```
Comment (Upabjojr) [inline sympy/stats/crv_types.py]: two spaces?

File: sympy/stats/crv_types.py [hunk 3]
```diff
@@ -1948,7 +1953,7 @@ def Normal(name, mean, std):
 
     mu : Real number or a list representing the mean or the mean vector
     sigma : Real number or a positive definite sqaure matrix,
-         :math:`\sigma^2 > 0` the variance
```
Comment (Upabjojr) [inline sympy/stats/crv_types.py]: sigma squares is the variance

File: sympy/stats/rv.py [hunk 4]
```diff
@@ -121,6 +121,27 @@ def as_boolean(self):
         return And(self.fulldomain.as_boolean(), self.condition)
 
 
+class RandomDistribution(Basic):
```
Comment (Upabjojr) [inline sympy/stats/rv.py]: what about `ProbabilityDistribution`?

File: sympy/stats/rv.py [hunk 5]
```diff
@@ -121,6 +121,27 @@ def as_boolean(self):
...
+    def handle_compound_dist(self, expr):
+        from sympy.stats.rv import density
+        from sympy.concrete.summations import summation
+        for arg in self.args:
+            if isinstance(arg, RandomSymbol):
+                sym, dom = arg.symbol, arg.pspace.set
+                lpdf = density(arg)(sym)
+                expr = expr.replace(arg, sym)*lpdf
+                if arg.pspace.is_Continuous:
+                    expr = integrate(expr, (sym, dom))
```
Comment (Upabjojr) [inline sympy/stats/rv.py]: `integrate` does not currently support the set notation, but it should in the future. Better leave a note.

File: sympy/stats/rv.py [hunk 6]
```diff
@@ -121,6 +121,27 @@ def as_boolean(self):
...
+                elif arg.pspace.is_Discrete:
+                    if dom in (S.Integers, S.Naturals, S.Naturals0):
```
Comment (Upabjojr) [inline sympy/stats/rv.py]: this should be transferred to `summation` or `Sum`. But better open an issue an leave it for another PR.

File: sympy/stats/tests/test_continuous_rv.py [hunk 7]
```diff
@@ -20,7 +20,7 @@
 
 
 from sympy.stats.crv_types import NormalDistribution
-from sympy.stats.rv import IndependentProductPSpace
+from sympy.stats.crv import  SingleContinuousPSpace
```
Comment (Upabjojr) [inline sympy/stats/tests/test_continuous_rv.py]: two spaces

**Comment by Upabjojr** [comment]
This PR looks good. It needs some more tests.

File: sympy/stats/crv.py (hunk 1 above)
Comment (akash9712) [reply sympy/stats/crv.py]: The difference is still present in other classes such as `SingleContinuousDistribution`, but `ContinuousDistribution` and `DiscreteDistribution` classes were identical, `__call__` being the only method described. Adding `handle_compound_dist` would also lead to some more code repitition.

File: sympy/stats/crv_types.py (hunk 2 above)
Comment (akash9712) [reply sympy/stats/crv_types.py]: Sorry, made a typo.

**Comment by Upabjojr** [comment]
```python
def test_random_parameters_given():
    mu = Normal('mu', 2, 3)
    meas = Normal('T', mu, 1)
    assert given(meas, Eq(mu, 5)) == Normal('T', 5, 1)
```
This test makes sense, we need to keep this behaviour.

**Comment by Upabjojr** [comment]
```python
def test_conjugate_priors():
    mu = Normal('mu', 2, 3)
    x = Normal('x', mu, 1)
    assert isinstance(simplify(density(mu, Eq(x, y), evaluate=False)(z)),
            Integral)
```
In this case, the integral gets calculated. Do you prefer to change the test or shall we keep the integral unevaluated?

File: sympy/stats/crv_types.py [hunk 8]
```diff
@@ -143,6 +143,15 @@ def rv(symbol, cls, args):
     args = list(map(sympify, args))
     dist = cls(*args)
     dist.check(*args)
+    if any([isinstance(arg, RandomSymbol) for arg in args]):
+        symbol = sympify(symbol)
+        new_dist = find_dist(cls, args)
+        if  new_dist != None:
+            return rv(symbol, new_dist[0], new_dist[1])
+        _pdf = dist.pdf(symbol)
+        _pdf, _set = dist.handle_compound_dist(_pdf), dist.set
```
Comment (Upabjojr) [inline sympy/stats/crv_types.py]: I would not calculate the integral at this level. I would recommend creating a marginalized joint distribution, without performing the integration.

**Comment by Upabjojr** [comment]
We have `JointDistribution`. I think we need something called `MarginalDistribution` as well... the pspace of a compound distribution could then be `(X, MarginalDistribution(JointDistribution(...), ... ))`.

What do you think?

**Comment by akash9712** [comment]
> We have `JointDistribution`. I think we need something called `MarginalDistribution` as well... the pspace of a compound distribution could then be `(X, MarginalDistribution(JointDistribution(...), ... ))`

I think this could work, and would help if a condition is imposed on the latent distribution.

**Comment by akash9712** [comment]
@Upabjojr I have created the class `MarginalDistribution`, but I am still working on how to fix the failing tests; I am trying to modify `given` so that the conditions on latent distributions can be imposed.

File: sympy/stats/crv_types.py [hunk 9]
```diff
@@ -143,7 +144,13 @@ def rv(symbol, cls, args):
     args = list(map(sympify, args))
     dist = cls(*args)
     dist.check(*args)
-    return SingleContinuousPSpace(symbol, dist).value
+    pspace = SingleContinuousPSpace(symbol, dist)
+    latent_dist = [arg for arg in args if isinstance(arg, RandomSymbol)]
+    _pdf = pspace.pdf
+    if latent_dist != []:
```
Comment (ArifAhmed1995) [inline sympy/stats/crv_types.py]: I guess a more pythonic way is to use `if latent_dist:`

File: sympy/stats/drv_types.py [hunk 10]
```diff
@@ -15,7 +16,13 @@ def rv(symbol, cls, *args):
     args = list(map(sympify, args))
     dist = cls(*args)
     dist.check(*args)
-    return SingleDiscretePSpace(symbol, dist).value
+    pspace = SingleDiscretePSpace(symbol, dist)
+    latent_dist = [arg for arg in args if isinstance(arg, RandomSymbol)]
+    _pdf = pspace.pdf
+    if latent_dist != []:
```
Comment (ArifAhmed1995) [inline sympy/stats/drv_types.py]: Same here...

File: sympy/stats/joint_rv.py [hunk 11]
```diff
@@ -163,3 +165,45 @@ def marginal_distribution(rv, *indices):
...
+
+    def check(self):
+        pass
+
+    def set(self):
+        rvs = (i for i in random_symbols(self.args[1]))
+        return ProductSet((i.pspace.set for i in rvs))
+
+    def pdf(self, x):
+        expr = self.args[1]
```
Comment (ArifAhmed1995) [inline sympy/stats/joint_rv.py]: It might be a good idea to extract all the used arguments into variables with proper names. Apart from `args[1]` which refers to the expression, I cannot make out what the other two refer to.
In future, it makes it difficult for other contributors to work and modify the code.

File: sympy/stats/rv.py [hunk 12]
```diff
@@ -517,6 +521,8 @@ def pspace(expr):
     True
     """
     expr = sympify(expr)
+    if isinstance(expr, RandomSymbol) and expr.pspace != None:
```
Comment (ArifAhmed1995) [inline sympy/stats/rv.py]: Try to use `expr.pspace is not None`.

**Comment by akash9712** [comment]
@ArifAhmed1995 The work done in this PR was included in #14989. Closing this now.
